Add Support for Excess Income Report

ivendor opened the pull request:
For offshore funds that reports to HMRC (all the usual Vanguard, Blackrock, XTrackers ETFs for example), there are additional taxes to pay beyond CGT. Full list at: https://www.gov.uk/government/publications/offshore-funds-list-of-reporting-funds

Since the fund accumulate capitals every year that do not re-distribute as dividends (this is true for both Accumulating and Distributing funds), in UK we need to pay taxes on such excess as income tax. This also means that accumulating funds are as tax efficient as distributing counterparts here in UK.

This type of taxation is a bit complex and is defined in: https://www.gov.uk/government/publications/offshore-funds-self-assessment-helpsheet-hs265/hs265-offshore-funds

Vanguard UK also has a simplified FAQ on this, with explain principles and calculation: https://www.vanguardinvestor.co.uk/content/dam/intl/uk-retail-direct/general/uk-reporting-fund-faq.pdf

Plus there are implications with Bed and Breakfast explained well here too:
https://www.rawknowledge.ltd/eri-explained-four-tricky-questions-answered/

KapJI commented on cgt_calc/parsers/eri/__init__.py:
```diff
@@ -0,0 +1,24 @@
+"""Parse ERI input files."""
```
Add more details here about what this is.

ivendor commented:
> So I just spent considerable time getting the data for Vanguard for 2022 to 2024 (before that they have ugly PDFs). This is hideous work on top of what we already need to do. To solve this, I propose:
> 
> 1. Instead of the `Symbol` column in the csv, we use `ISIN` which is unique. This way the ERI file should be useful to everybody.
> 2. We can then bundle the ERI in the package. We can create one ERI file per fund source, then load it automatically. Of course it should still be possible to override it.
> 3. Some brokers provide the ISIN already (check Trading212 parser), we can make it work automatically when that's available
> 4. For the brokers that don't, we can provide a way to indicate the Symbol -> ISIN mapping, even with some defaults. This way the user might be able to get ERI pretty much out-of-the-box.
> 
> This is the data I compiled for Vanguard, so we don't need to do it again: [vanguard_eri.csv](https://github.com/user-attachments/files/22671241/vanguard_eri.csv)

yeah good points, about ISIN translation we could try to use Bloomberg open APIs and storing it in a local cache similar to exchange rates:

```
curl 'https://api.openfigi.com/v3/mapping' --request POST --header 'Content-Type: application/json' --data '[{"idType":"ID_ISIN","idValue":"IE000EKJRSZ3"}]'
```

ivendor commented:
@vmartinv  at the moment I can't use your file as is because it will require ticker translations for all the tickers so for now I won't include it. Once we add automated ticker translations we can add the file.

vmartinv commented:
Wouldn’t it be easier if we add the ISIN fetching class first and then come back to this PR?

ivendor commented:
See https://github.com/KapJI/capital-gains-calculator/pull/606

ivendor commented on cgt_calc/main.py:
```diff
@@ -427,8 +540,9 @@ def convert_to_hmrc_transactions(
                 )
-                if self.date_in_tax_year(transaction.date):
-                    dividends[(symbol, currency)] += amount
+                _accumulate_if_in_tax_year(
+                    transaction.date, dividends, (symbol, currency), amount
+                )
```
I can't otherwise the lint won't accept the function (too much cyclomatic complexity)

ivendor replied:
That causes unnecessary printing of dividends and interests even with 0s

vmartinv replied:
looks good now!

vmartinv commented on cgt_calc/model.py:
```diff
@@ -378,6 +440,21 @@ def __str__(self) -> str:
+                RuleType.EXCESS_REPORTED_INCOME_DISTRIBUTION,
+            )
+        )
+        if eris:
+            out += "ERI distributed:\n"
+            for item in self._filter_calculation_log(
+                self.calculation_log_yields,
+                RuleType.EXCESS_REPORTED_INCOME_DISTRIBUTION,
+            ):
+                assert item.eri
+                out += f"  {item.eri.symbol}: £{round_decimal(item.amount,2)}\n"
+
```
This should be more clear
```
Excess Reportable Income distributed:
  VUAG: £22.26 (counts as interest)
  VNRG: £0.91 (counts as dividends)
```

ivendor commented:
Addressed comments, other note for the future, the foreign interests and dividends has to be reported by country, we can use the ISIN country code to properly do so in the report

vmartinv commented on tests/test_data/test_run_with_example_files_output.txt:
```diff
@@ -18,25 +19,28 @@ Dividends:
   NVDA: 1.00, £401.01
-  FOO: 30.50, £543.97
   GOOG: 172.00, £12516.55
+Excess Reported Income:
+  VUAG: £22.26 (included as dividend)
+  VNRG: £0.91 (included as dividend)
 For tax year 2020/2021:
```
One last thing, move ERI down to reflect that is part of the current tax year's calculations.
```suggestion
For tax year 2020/2021:
Excess Reported Income:
  VUAG: £22.26 (included as dividend)
  VNRG: £0.91 (included as dividend)
```

vmartinv commented on README.md:
```diff
@@ -171,6 +174,103 @@ Example usage for the tax year 2024/25:
+For correct taxation on [offshore funds](https://www.gov.uk/government/publications/offshore-funds-self-assessment-helpsheet-hs265/hs265-offshore-funds) you need to specify the yearly excess reported income (ERI) from each fund you have owned.
+You can find the full list of funds that requires this at [HMRC](https://www.gov.uk/government/publications/offshore-funds-list-of-reporting-funds).
+
+The tool already includes such yearly history in the [resources folder](https://github.com/KapJI/capital-gains-calculator/blob/main/cgt_calc/resources/eri). You can check if you're fund is already included, or provide a custom ERI history file using the instructions below.
```
```suggestion
The tool already includes such yearly history in the [resources folder](https://github.com/KapJI/capital-gains-calculator/blob/main/cgt_calc/resources/eri). You can check if your fund is already included, or provide a custom ERI history file following the instructions [here](https://github.com/KapJI/capital-gains-calculator/blob/main/excess_reported_income_sources.md)). We strongly suggest sharing compiled ERI data so it can be added to the package as it can save significant time to other users that hold the same fund.
```

vmartinv commented on README.md:
```diff
@@ -171,6 +174,103 @@ Example usage for the tax year 2024/25:
+
+The ERI funds are indexed by ISIN and the tool provides automatic translation from ISIN to tickers, in case your broker doesn't supply the ISIN in their transaction history. For instructions on how to override ISIN translation look at the Extra options section below.
+
+There are few **unsupported** functionality at the moment for taxation on offshore funds:
```
```suggestion
There are a few **unsupported** functionalities at the moment for taxation on offshore funds:
```

vmartinv commented on README.md:
```diff
@@ -171,6 +174,103 @@ Example usage for the tax year 2024/25:
+They are split in Index Funds (BGIF), Global Funds (BGF), Strategic Funds (BSF)
+
+Columns mapping to ERI_RAW:
+
+-   **ISIN:** same name column
+-   **Fund Reporting Period End Date:** End date in the Reporting Period column
+-   **Currency:** same name column
+-   **Excess of reporting income over distribution:** Excess of reporting income per unit column
+</details>
+ <br />
+<details>
+    <summary>🏦 Instructions for Xtrackers funds</summary>
+
+DWS UK publishes the Reportable Income yearly report at the bottom of this page:
+https://etf.dws.com/en-gb/information/etf-documents/reportings/
+
+They are split XTrackers (stocks ETF), XTrackers II (bonds ETF) and XTrackers IE (other stocks ETF).
+
+Columns mapping to ERI_RAW:
+
+-   **ISIN:** same name column
+-   **Fund Reporting Period End Date:** Period Ended date at the top of the PDF
+-   **Currency:** Share class currency column
+-   **Excess of reporting income over distribution:** Excess reported income per share column
+</details>
+ <br />
+<details>
+    <summary>🏦 Instructions for Amundi funds</summary>
+
+Amundi UK publishes the Reportable Income yearly report at the bottom of this page:
+https://www.amundietf.co.uk/en/individual/resources/document-library?documentType=uktaxcalculation
+
+They are split XTrackers (stocks ETF), XTrackers II (bonds ETF) and XTrackers IE (other stocks ETF).
+
+Columns mapping to ERI_RAW:
+
+-   **ISIN:** same name column
+-   **Fund Reporting Period End Date:** Reporting Period End Date column
+-   **Currency:** Currency of the following amounts column
+-   **Excess of reporting income over distribution:** Per unit excess reportable income over distributions in respect of the reporting period column
```
move to `excess_reported_income_sources.md`

vmartinv reviewed (approved):
Code looks great now. I leave @KapJI to take another look. This is a big PR but @ivendor  managed to fix this big gap we had in the calculations with very little friction to the user. I’m happy to say that for my accounts I didn’t have to do anything to make it work, it just calculated it from the bundled data. Thank you @ivendor !

KapJI commented:
@ivendor Can you fix lint warning?

KapJI commented on pyproject.toml:
```diff
@@ -33,6 +33,7 @@ requests = "^2.27.1"
 types-requests = "^2.27.7"
 yfinance = "^0.2.36"
 requests-ratelimiter = "^0.7.0"
+types-python-dateutil = "^2.9.0.20250822"
```
This should go to dev dependencies. Together with `types-requests`.

KapJI commented on excess_reported_income_sources.md:
````diff
@@ -0,0 +1,84 @@
+Example usage for the tax year 2024/25:
+
+```shell
+cgt-calc --year 2024 --eri-raw-file eri_raw.csv [broker_transactions_options...]
````
Add a comment that for Vanguard that file is loaded automatically?

vmartinv replied:
Yeah we should add at the top of the file that this guide is needed when the fund is not already bundled. I switched between several providers so I will be adding later more coverage. We will have not just Vanguard.

KapJI commented on excess_reported_income_sources.md:
```diff
@@ -0,0 +1,84 @@
+You will need:
+
+-   **CSV using the ERI_RAW format.** This is currently the only format supported for excess reported income.
+    [See example](https://github.com/KapJI/capital-gains-calculator/blob/main/cgt_calc/resources/eri/vanguard_eri.csv)
```
```suggestion
    [See example.](https://github.com/KapJI/capital-gains-calculator/blob/main/cgt_calc/resources/eri/vanguard_eri.csv)
```
